File: 2022/day20.cpp
// https://adventofcode.com/2022/day/20

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <print>
#include <ranges>

using Input = Vec<i64>;

fn parse(String const& filename) -> Input {
  return aoc::views::read_numbers<i64>(filename) | aoc::ranges::to<Input>();
}

struct IndexedNum {
  isize index;
  i64 value;
};

template <usize num_mixings, class OGList>
fn reorder_list(OGList&& original_list) -> Vec<i64> {
  let numbers = std::forward<OGList>(original_list) |
                stdv::enumerate |
                stdv::transform([](auto pair) -> IndexedNum {
                  let[i, v] = pair;
                  return {i, v};
                }) |
                aoc::collect_vec<IndexedNum>();
  auto mixed = numbers;

  for (let _ : Range{0uz, num_mixings}) {
    for (let& indexed_number : numbers) {
      let[orig_index, orig_value] = indexed_number;
      if (orig_value == 0) {
        continue;
      }

      let current_it = stdr::find(mixed, orig_index, &IndexedNum::index);

      let next_it = [&] {
        let begin = stdr::begin(mixed);
        let size = std::ssize(mixed) - 1;
        auto pos = std::distance(begin, current_it);
        pos = (size + (pos + (orig_value % size))) % size;
        return begin + pos;
      }();

      let it_distance = std::distance(current_it, next_it);
      if (it_distance < 0) {
        std::shift_right(next_it, current_it + 1, 1);
      } else {
        std::shift_left(current_it, next_it + 1, 1);
      }
      *next_it = indexed_number;
    }
  }

  return mixed | stdv::transform(&IndexedNum::value) | aoc::collect_vec<i64>();
}

fn sum_numbers(Vec<i64> const& list) -> i64 {
  let n = std::ssize(list);
  let zero_pos = stdr::distance(stdr::begin(list), stdr::find(list, i64{0}));
  let at = [&](isize offset) -> i64 {
    return list[static_cast<usize>((zero_pos + offset) % n)];
  };
  return at(1000) + at(2000) + at(3000);
}

template <usize num_mixings, i64 decryption_key>
fn solve_case(Input const& input) -> i64 {
  let mixed = reorder_list<num_mixings>(
      input | stdv::transform([](i64 v) { return v * decryption_key; }));
  return sum_numbers(mixed);
}

int main() {
  std::println("Part 1");
  let example = parse("day20.example");
  AOC_EXPECT_RESULT(3, (solve_case<1, 1>(example)));
  let input = parse("day20.input");
  AOC_EXPECT_RESULT(1591, (solve_case<1, 1>(input)));

  std::println("Part 2");
  AOC_EXPECT_RESULT(1623178306, (solve_case<10, 811589153>(example)));
  AOC_EXPECT_RESULT(14579387544492, (solve_case<10, 811589153>(input)));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day20.rs
fn parse(filename: &str) -> Vec<i64> {
    aoc::file::read_numbers(filename)
}

fn solve_case<const NUM_MIXES: usize, const DECRYPTION_KEY: i64>(numbers: &[i64]) -> i64 {
    let numbers = numbers
        .iter()
        .map(|num| num * DECRYPTION_KEY)
        .enumerate()
        .collect::<Vec<_>>();
    let mut mixed = numbers.clone();

    for _ in 0..NUM_MIXES {
        for &(orig_index, orig_value) in numbers.iter() {
            if (orig_value == 0) {
                continue;
            }
            let current_ipos = mixed
                .iter()
                .position(|(index, _)| *index == orig_index)
                .unwrap() as i64;
            let next_ipos = if (orig_value == 0) {
                current_ipos
            } else {
                // Very important to subtract one
                (current_ipos + orig_value).rem_euclid(mixed.len() as i64 - 1)
            };
            let distance = next_ipos - current_ipos;
            let current_pos = current_ipos as usize;
            let next_pos = next_ipos as usize;
            if (distance < 0) {
                // shift_right
                mixed.copy_within(next_pos..current_pos, next_pos + 1);
            } else {
                // shift_left
                mixed.copy_within(current_pos + 1..next_pos + 1, current_pos);
            }
            mixed[next_ipos as usize] = (orig_index, orig_value);
        }
    }

    let mut pos = mixed.iter().position(|(_, value)| *value == 0).unwrap();
    return (0..3)
        .map(|_| {
            pos = (pos + 1000) % mixed.len();
            return mixed[pos].1;
        })
        .sum();
}

fn main() {
    println!("Part 1");
    let example = parse("day20.example");
    assert_eq!(3, solve_case::<1, 1>(&example));
    let input = parse("day20.input");
    assert_eq!(1591, solve_case::<1, 1>(&input));

    println!("Part 2");
    assert_eq!(1623178306, solve_case::<10, 811589153>(&example));
    assert_eq!(14579387544492, solve_case::<10, 811589153>(&input));
}
